types/wlr_idle.c

```diff
@@ -246,3 +216,65 @@ struct wlr_idle *wlr_idle_create(struct wl_display *display) {
 void wlr_idle_notify_activity(struct wlr_idle *idle, struct wlr_seat *seat) {
 	wlr_signal_emit_safe(&idle->events.activity_notify, seat);
 }
+
+struct wlr_idle_timeout *wlr_idle_timeout_create(struct wlr_idle *idle,
+		struct wlr_seat *seat, uint32_t timeout) {
+	struct wlr_idle_timeout *timer =
+		calloc(1, sizeof(struct wlr_idle_timeout));
+	if (!timer) {
+		return NULL;
+	}
+
+	timer->seat = seat;
+	timer->timeout = timeout;
+	timer->idle_state = false;
+	timer->enabled = idle->enabled;
+
+	wl_list_insert(&idle->idle_timers, &timer->link);
+	wl_signal_init(&timer->events.idle);
+	wl_signal_init(&timer->events.resume);
+	wl_signal_init(&timer->events.destroy);
+
+	timer->seat_destroy.notify = handle_seat_destroy;
+	wl_signal_add(&timer->seat->events.destroy, &timer->seat_destroy);
+
+	timer->input_listener.notify = handle_input_notification;
+	wl_signal_add(&idle->events.activity_notify, &timer->input_listener);
+	// create the timer
+	timer->idle_source =
+		wl_event_loop_add_timer(idle->event_loop, idle_notify, timer);
+	if (timer->idle_source == NULL) {
+		wl_list_remove(&timer->link);
+		wl_list_remove(&timer->input_listener.link);
+		wl_list_remove(&timer->seat_destroy.link);
+		wl_resource_set_user_data(timer->resource, NULL);
+		free(timer);
+
+		return NULL;
+	}
+
+	if (timer->enabled) {
+		// arm the timer
+		wl_event_source_timer_update(timer->idle_source, timer->timeout);
+		if (timer->timeout == 0) {
+			idle_notify(timer);
```

Comment:
This won't work, `timer->resource` is NULL at this point.

I guess the solution will be to keep a private constructor with a `resource` argument, and expose a public constructor which provides `resource = NULL` to the private constructor.